// breeze-client/esm2015/src/unattached-children-map.js
import { core } from './core';
/** @hidden @internal */
// Represents entities not yet attached to navigationProperties. 
export class UnattachedChildrenMap {
    constructor() {
        // key is EntityKey.toString(), value is array of { navigationProperty, children }
        this.map = {};
    }
    addChild(parentEntityKey, navigationProperty, child) {
        let tuple = this.getTuple(parentEntityKey, navigationProperty);
        if (!tuple) {
            tuple = { navigationProperty: navigationProperty, children: [] };
            core.getArray(this.map, parentEntityKey.toString()).push(tuple);
        }
        tuple.children.push(child);
    }
    removeChildren(parentEntityKeyString, navigationProperty) {
        let tuples = this.map[parentEntityKeyString];
        if (!tuples)
            return;
        core.arrayRemoveItem(tuples, (t) => {
            return t.navigationProperty === navigationProperty;
        });
        if (!tuples.length) {
            delete this.map[parentEntityKeyString];
        }
    }
    getTuple(parentEntityKey, navigationProperty) {
        let tuples = this.getTuples(parentEntityKey);
        if (!tuples)
            return null;
        let tuple = core.arrayFirst(tuples, function (t) {
            return t.navigationProperty === navigationProperty;
        });
        return tuple;
    }
    getTuples(parentEntityKey) {
        let allTuples = [];
        let tuples = this.map[parentEntityKey.toString()];
        if (tuples) {
            allTuples = allTuples.concat(tuples);
        }
        let entityType = parentEntityKey.entityType;
        while (entityType.baseEntityType) {
            entityType = entityType.baseEntityType;
            let baseKey = parentEntityKey.toString(entityType);
            tuples = this.map[baseKey];
            if (tuples) {
                allTuples = allTuples.concat(tuples);
            }
        }
        return (allTuples.length) ? allTuples : undefined;
    }
    getTuplesByString(parentEntityKeyString) {
        return this.map[parentEntityKeyString];
    }
}
