// breeze-client/esm5/src/complex-array.js
import { core } from './core';
import { observableArray } from './observable-array';
import { BreezeEvent } from './event';
var ɵ0 = function (adds) {
    return getGoodAdds(this, adds);
}, ɵ1 = function () {
    observableArray.updateEntityState(this);
}, ɵ2 = function (adds) {
    processAdds(this, adds);
}, ɵ3 = function (removes) {
    processRemoves(this, removes);
}, ɵ4 = function () {
    if (!this._origValues)
        return;
    var that = this;
    this.forEach(function (co) {
        clearAspect(co, that);
    });
    this.length = 0;
    this._origValues.forEach(function (co) {
        that.push(co);
    });
}, ɵ5 = function () {
    this._origValues = null;
};
var complexArrayMixin = {
    // complexArray will have the following props
    //    parent
    //    propertyPath
    //    parentProperty
    //    addedItems  - only if modified
    //    removedItems  - only if modified
    //  each complexAspect of any entity within a complexArray
    //  will have its own _complexState = "A/M";
    /**
     Complex arrays are not actually classes, they are objects that mimic arrays. A complex array is collection of
     complexTypes associated with a data property on a single entity or other complex object. i.e. customer.orders or order.orderDetails.
     This collection looks like an array in that the basic methods on arrays such as 'push', 'pop', 'shift', 'unshift', 'splice'
     are all provided as well as several special purpose methods.
     @class {complexArray}
     **/
    /**
    An [[Event]] that fires whenever the contents of this array changed.  This event
    is fired any time a new entity is attached or added to the EntityManager and happens to belong to this collection.
    Adds that occur as a result of query or import operations are batched so that all of the adds or removes to any individual
    collections are collected into a single notification event for each relation array.
    @example
        // assume order is an order entity attached to an EntityManager.
        orders.arrayChanged.subscribe(
        function (arrayChangedArgs) {
            var addedEntities = arrayChangedArgs.added;
            var removedEntities = arrayChanged.removed;
        });
    @event arrayChanged
    @param added {Array of Entity} An array of all of the entities added to this collection.
    @param removed {Array of Entity} An array of all of the removed from this collection.
    @readOnly
    **/
    // virtual impls
    _getGoodAdds: ɵ0,
    _beforeChange: ɵ1,
    _processAdds: ɵ2,
    _processRemoves: ɵ3,
    _rejectChanges: ɵ4,
    _acceptChanges: ɵ5
};
// local functions
function getGoodAdds(complexArray, adds) {
    // remove any that are already added here
    return adds.filter(function (a) {
        // return a.parent !== complexArray.parent;  // TODO: check if this is actually a bug in original breezejs ???
        return a.complexAspect == null || a.complexAspect.parent !== complexArray.parent;
    });
}
function processAdds(complexArray, adds) {
    adds.forEach(function (a) {
        // if (a.parent != null) { // TODO: check if this is actually a bug in original breezejs ???
        if (a.complexAspect && a.complexAspect.parent != null) {
            throw new Error("The complexObject is already attached. Either clone it or remove it from its current owner");
        }
        setAspect(a, complexArray);
    });
}
function processRemoves(complexArray, removes) {
    removes.forEach(function (a) {
        clearAspect(a, complexArray);
    });
}
function clearAspect(co, arr) {
    var coAspect = co.complexAspect;
    // if not already attached - exit
    if (coAspect.parent !== arr.parent)
        return null;
    coAspect.parent = undefined;
    coAspect.parentProperty = undefined;
    return coAspect;
}
function setAspect(co, arr) {
    var coAspect = co.complexAspect;
    // if already attached - exit
    if (coAspect.parent === arr.parent)
        return null;
    coAspect.parent = arr.parent;
    coAspect.parentProperty = arr.parentProperty;
    return coAspect;
}
/** For use by breeze plugin authors only. The class is for use in building a [[IModelLibraryAdapter]] implementation.
@adapter (see [[IModelLibraryAdapter]])
@hidden
*/
export function makeComplexArray(arr, parent, parentProperty) {
    var arrX = arr;
    observableArray.initializeParent(arrX, parent, parentProperty);
    arrX.arrayChanged = new BreezeEvent("arrayChanged", arrX);
    core.extend(arrX, observableArray.mixin);
    return core.extend(arrX, complexArrayMixin);
}
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5 };
