// breeze-client/esm5/src/config.js
import { core } from './core';
import { assertParam } from './assert-param';
import { BreezeEvent } from './event';
var InterfaceDef = /** @class */ (function () {
    function InterfaceDef(name) {
        this.name = name;
        this.defaultInstance = undefined;
        this._implMap = {};
    }
    /** Define an implementation of the given adaptername */
    InterfaceDef.prototype.registerCtor = function (adapterName, ctor) {
        this._implMap[adapterName.toLowerCase()] = { ctor: ctor, defaultInstance: undefined };
    };
    /** Return the definition for the given adapterName */
    InterfaceDef.prototype.getImpl = function (adapterName) {
        return this._implMap[adapterName.toLowerCase()];
    };
    /** Return the first implementation for this InterfaceDef */
    InterfaceDef.prototype.getFirstImpl = function () {
        var kv = core.objectFirst(this._implMap, function () {
            return true;
        });
        return kv ? kv.value : null;
    };
    InterfaceDef.prototype.getDefaultInstance = function () {
        return this.defaultInstance;
    };
    return InterfaceDef;
}());
export { InterfaceDef };
var BreezeConfig = /** @class */ (function () {
    function BreezeConfig() {
        this.functionRegistry = {};
        this.typeRegistry = {};
        this.objectRegistry = {};
        this.stringifyPad = '';
        this.interfaceInitialized = new BreezeEvent("interfaceInitialized", this);
        if (this.noEval === undefined) {
            try {
                Error['x'] = Function('');
                this.noEval = false; // eval succeeded
            }
            catch (_a) {
                this.noEval = true; // eval failed, probably due to CSP
            }
        }
    }
    /**
    Method use to register implementations of standard breeze interfaces.  Calls to this method are usually
    made as the last step within an adapter implementation.
    @method registerAdapter
    @param interfaceName {String} - one of the following interface names: "ajax", "dataService", "modelLibrary", "uriBuilder"
    @param adapterCtor {Function} - an ctor function that returns an instance of the specified interface.
    **/
    BreezeConfig.prototype.registerAdapter = function (interfaceName, adapterCtor) {
        assertParam(interfaceName, "interfaceName").isNonEmptyString().check();
        assertParam(adapterCtor, "adapterCtor").isFunction().check();
        // this impl will be thrown away after the name is retrieved.
        var impl = new adapterCtor();
        var implName = impl.name;
        if (!implName) {
            throw new Error("Unable to locate a 'name' property on the constructor passed into the 'registerAdapter' call.");
        }
        var idef = this.getInterfaceDef(interfaceName);
        idef.registerCtor(implName, adapterCtor);
    };
    /**
    Returns the ctor function used to implement a specific interface with a specific adapter name.
    @method getAdapter
    @param interfaceName {String} One of the following interface names: "ajax", "dataService", "modelLibrary", "uriBuilder"
    @param [adapterName] {String} The name of any previously registered adapter. If this parameter is omitted then
    this method returns the "default" adapter for this interface. If there is no default adapter, then a null is returned.
    @return {Function|null} Returns either a ctor function or null.
    **/
    BreezeConfig.prototype.getAdapter = function (interfaceName, adapterName) {
        var idef = this.getInterfaceDef(interfaceName);
        if (adapterName) {
            var impl = idef.getImpl(adapterName);
            return impl ? impl.ctor : null;
        }
        else {
            return idef.defaultInstance ? idef.defaultInstance._$impl.ctor : null;
        }
    };
    /**
    Initializes a single adapter implementation. Initialization means either newing a instance of the
    specified interface and then calling "initialize" on it or simply calling "initialize" on the instance
    if it already exists.
    @method initializeAdapterInstance
    @param interfaceName {String} The name of the interface to which the adapter to initialize belongs.
    @param adapterName {String} - The name of a previously registered adapter to initialize.
    @param [isDefault=true] {Boolean} - Whether to make this the default "adapter" for this interface.
    @return {an instance of the specified adapter}
    **/
    BreezeConfig.prototype.initializeAdapterInstance = function (interfaceName, adapterName, isDefault) {
        if (isDefault === void 0) { isDefault = true; }
        isDefault = isDefault === undefined ? true : isDefault;
        assertParam(interfaceName, "interfaceName").isNonEmptyString().check();
        assertParam(adapterName, "adapterName").isNonEmptyString().check();
        assertParam(isDefault, "isDefault").isBoolean().check();
        var idef = this.getInterfaceDef(interfaceName);
        var impl = idef.getImpl(adapterName);
        if (!impl) {
            throw new Error("Unregistered adapter.  Interface: " + interfaceName + " AdapterName: " + adapterName);
        }
        return this._initializeAdapterInstanceCore(idef, impl, isDefault);
    };
    /**
    Returns the adapter instance corresponding to the specified interface and adapter names.
    @method getAdapterInstance
    @param interfaceName {String} The name of the interface.
    @param [adapterName] {String} - The name of a previously registered adapter.  If this parameter is
    omitted then the default implementation of the specified interface is returned. If there is
    no defaultInstance of this interface, then the first registered instance of this interface is returned.
    @return {an instance of the specified adapter}
    @internal
    **/
    BreezeConfig.prototype.getAdapterInstance = function (interfaceName, adapterName) {
        var idef = this.getInterfaceDef(interfaceName);
        var impl;
        var isDefault = adapterName == null || adapterName === "";
        if (isDefault) {
            if (idef.defaultInstance)
                return idef.defaultInstance;
            impl = idef.getFirstImpl();
        }
        else {
            impl = idef.getImpl(adapterName);
        }
        if (!impl)
            return undefined;
        if (impl.defaultInstance) {
            return impl.defaultInstance;
        }
        else {
            return this._initializeAdapterInstanceCore(idef, impl, isDefault);
        }
    };
    /** this is needed for reflection purposes when deserializing an object that needs a fn or ctor.
        Used to register validators. */
    BreezeConfig.prototype.registerFunction = function (fn, fnName) {
        assertParam(fn, "fn").isFunction().check();
        assertParam(fnName, "fnName").isString().check();
        if (fn.prototype) {
            fn.prototype._$fnName = fnName;
        }
        this.functionRegistry[fnName] = fn;
    };
    BreezeConfig.prototype.registerType = function (ctor, typeName) {
        assertParam(ctor, "ctor").isFunction().check();
        assertParam(typeName, "typeName").isString().check();
        if (ctor.prototype) {
            ctor.prototype._$typeName = typeName;
        }
        this.typeRegistry[typeName] = ctor;
    };
    BreezeConfig.prototype.getRegisteredFunction = function (fnName) {
        return this.functionRegistry[fnName];
    };
    BreezeConfig.prototype.getInterfaceDef = function (interfaceName) {
        var lcName = interfaceName.toLowerCase();
        // source may be null
        var kv = core.objectFirst(this._interfaceRegistry || {}, function (k, v) {
            return k.toLowerCase() === lcName;
        });
        if (!kv) {
            throw new Error("Unknown interface name: " + interfaceName);
        }
        return kv.value;
    };
    /** @deprecated @internal no-op kept for backward compatibility */
    BreezeConfig.prototype.setQ = function (q) {
        console && console.warn("setQ does nothing; ES6 Promise support is required - use a shim if necessary.");
    };
    /** @hidden @internal */
    BreezeConfig.prototype._storeObject = function (obj, type, name) {
        // uncomment this if we make this public.
        //assertParam(obj, "obj").isObject().check();
        //assertParam(name, "objName").isString().check();
        var key = (typeof (type) === "string" ? type : type.prototype._$typeName) + "." + name;
        this.objectRegistry[key] = obj;
    };
    /** @hidden @internal */
    BreezeConfig.prototype._fetchObject = function (type, name) {
        if (!name)
            return undefined;
        var key = (typeof (type) === "string" ? type : type.prototype._$typeName) + "." + name;
        var result = this.objectRegistry[key];
        if (!result) {
            throw new Error("Unable to locate a registered object by the name: " + key);
        }
        return result;
    };
    /** @hidden @internal */
    BreezeConfig.prototype._initializeAdapterInstanceCore = function (interfaceDef, impl, isDefault) {
        var instance;
        var inst = impl.defaultInstance;
        if (!inst) {
            instance = new (impl.ctor)();
            impl.defaultInstance = instance;
            instance._$impl = impl;
        }
        else {
            instance = inst;
        }
        instance.initialize();
        if (isDefault) {
            // next line needs to occur before any recomposition
            interfaceDef.defaultInstance = instance;
        }
        // recomposition of other impls will occur here.
        this.interfaceInitialized.publish({ interfaceName: interfaceDef.name, instance: instance, isDefault: true });
        if (instance.checkForRecomposition != null) {
            // now register for own dependencies.
            this.interfaceInitialized.subscribe(function (interfaceInitializedArgs) {
                // TODO: why '!'s needed here for typescript to compile correctly???
                instance.checkForRecomposition(interfaceInitializedArgs);
            });
        }
        return instance;
    };
    return BreezeConfig;
}());
export { BreezeConfig };
export var config = new BreezeConfig();
// legacy
core.config = config;
