// breeze-client/esm5/src/unattached-children-map.js
import { core } from './core';
/** @hidden @internal */
// Represents entities not yet attached to navigationProperties. 
var UnattachedChildrenMap = /** @class */ (function () {
    function UnattachedChildrenMap() {
        // key is EntityKey.toString(), value is array of { navigationProperty, children }
        this.map = {};
    }
    UnattachedChildrenMap.prototype.addChild = function (parentEntityKey, navigationProperty, child) {
        var tuple = this.getTuple(parentEntityKey, navigationProperty);
        if (!tuple) {
            tuple = { navigationProperty: navigationProperty, children: [] };
            core.getArray(this.map, parentEntityKey.toString()).push(tuple);
        }
        tuple.children.push(child);
    };
    UnattachedChildrenMap.prototype.removeChildren = function (parentEntityKeyString, navigationProperty) {
        var tuples = this.map[parentEntityKeyString];
        if (!tuples)
            return;
        core.arrayRemoveItem(tuples, function (t) {
            return t.navigationProperty === navigationProperty;
        });
        if (!tuples.length) {
            delete this.map[parentEntityKeyString];
        }
    };
    UnattachedChildrenMap.prototype.getTuple = function (parentEntityKey, navigationProperty) {
        var tuples = this.getTuples(parentEntityKey);
        if (!tuples)
            return null;
        var tuple = core.arrayFirst(tuples, function (t) {
            return t.navigationProperty === navigationProperty;
        });
        return tuple;
    };
    UnattachedChildrenMap.prototype.getTuples = function (parentEntityKey) {
        var allTuples = [];
        var tuples = this.map[parentEntityKey.toString()];
        if (tuples) {
            allTuples = allTuples.concat(tuples);
        }
        var entityType = parentEntityKey.entityType;
        while (entityType.baseEntityType) {
            entityType = entityType.baseEntityType;
            var baseKey = parentEntityKey.toString(entityType);
            tuples = this.map[baseKey];
            if (tuples) {
                allTuples = allTuples.concat(tuples);
            }
        }
        return (allTuples.length) ? allTuples : undefined;
    };
    UnattachedChildrenMap.prototype.getTuplesByString = function (parentEntityKeyString) {
        return this.map[parentEntityKeyString];
    };
    return UnattachedChildrenMap;
}());
export { UnattachedChildrenMap };
