// breeze-client/esm5/src/validation-options.js
import { core } from './core';
import { assertConfig } from './assert-param';
/**
A ValidationOptions instance is used to specify the conditions under which validation will be executed.

*/
var ValidationOptions = /** @class */ (function () {
    /**
    ValidationOptions constructor
    >     var newVo = new ValidationOptions( { validateOnSave: false, validateOnAttach: false });
    >     // assume em1 is a preexisting EntityManager
    >     em1.setProperties( { validationOptions: newVo });
    @param config - A configuration object.
    **/
    function ValidationOptions(config) {
        updateWithConfig(this, config);
    }
    /**
    Returns a copy of this ValidationOptions with changes to the specified config properties.
    >     var validationOptions = new ValidationOptions();
    >     var newOptions = validationOptions.using( { validateOnQuery: true, validateOnSave: false} );
    @param config - A configuration object
    @return A new ValidationOptions instance.
    **/
    ValidationOptions.prototype.using = function (config) {
        if (!config)
            return this;
        var result = new ValidationOptions(this);
        updateWithConfig(result, config);
        return result;
    };
    /**
    Sets the 'defaultInstance' by creating a copy of the current 'defaultInstance' and then applying all of the properties of the current instance.
    The current instance is returned unchanged.
    >     var validationOptions = new ValidationOptions()
    >     var newOptions = validationOptions.using( { validateOnQuery: true, validateOnSave: false} );
    >     var newOptions.setAsDefault();
    **/
    ValidationOptions.prototype.setAsDefault = function () {
        return core.setAsDefault(this, ValidationOptions);
    };
    /**
    The default instance for use whenever ValidationOptions are not specified.
    **/
    ValidationOptions.defaultInstance = new ValidationOptions({
        validateOnAttach: true,
        validateOnSave: true,
        validateOnQuery: false,
        validateOnPropertyChange: true
    });
    return ValidationOptions;
}());
export { ValidationOptions };
ValidationOptions.prototype._$typeName = "ValidationOptions";
function updateWithConfig(options, config) {
    if (config) {
        assertConfig(config)
            .whereParam("validateOnAttach").isBoolean().isOptional()
            .whereParam("validateOnSave").isBoolean().isOptional()
            .whereParam("validateOnQuery").isBoolean().isOptional()
            .whereParam("validateOnPropertyChange").isBoolean().isOptional()
            .applyAll(options);
    }
    return options;
}
